### PR TITLE
Add ability to use custom_target with `link_with` (fixes #4908)

### LINKED ISSUE #4908: Link against custom targets
For integration with custom toolchains and other such stuff. Basically if you put the output of custom_target to the `link_with` line of a build target it will:

 - add a dependency in the usual way
 - add the output file to the link line
 - assumes that this is enough to make it work, if not, then that is the build definition writer's developer's problem

This can probably only be done reasonably for static libraries. Dependencies of dependencies also don't really work, because to do so would mean having to replicate Meson's entire dependency resolution chain for custom targets.

I was going to do this today but am mostly been in bed suffering from the seasonal flu. Feel free to implement this if you wish.

### PR DESCRIPTION
**Pull request by TheQwertiest**
No docs yet - will add them if/when PR is considered valid.

~~Ideally, `CustomTargetIndex` (`custom_target[i]`) should be also supported, but it requires much more work, since `CustomTargetIndex` does not inherit from `Target`.~~
[EDIT]: Added `CustomTargetIndex` support.

### CODE REVIEW

**Comment by nirbheek** [comment]
I'm not sure I like the approach in this PR. It's basically doing duck-typing, which can be confusing. I think a cleaner way would be to create a new 'proxy' class that can encapsulate `CustomTarget` and `CustomTargetIndex`, and is automatically created when you pass one of those to `link_with:`.

**Comment by TheQwertiest** [comment]
Fair enough. Will be amended :)

File: mesonbuild/backend/vs2010backend.py [hunk 1]
```diff
@@ -1105,6 +1105,9 @@ def gen_vcxproj(self, target, ofname, guid):
 
         # Add more libraries to be linked if needed
         for t in target.get_dependencies():
+            if isinstance(t, CustomTarget) or isinstance(t, CustomTargetIndex):
```
Comment (ignatenkobrain) [inline mesonbuild/backend/vs2010backend.py]: __[Flake8]__ [F821] undefined name 'CustomTarget' ([link](https://sider.review/gh/repos/19784232/pulls/5066#issue-3834074))

<sub>posted by Sider</sub>

**Comment by TheQwertiest** [comment]
Personally, I don't like the workaround I've made in vs backend, so any suggestions are welcome.

**Comment by TheQwertiest** [comment]
@nirbheek Not sure that *Proxy* class idea pans out - everything that is passed to `link_with`/`link_whole` (and further down) must be a `Target` (in terms of `build.py`). Thus the Proxy class (`CustomTargetLinkTarget` in my case) must be a `Target` as well, which means introducing a new `Target` type to the backend, and that brings a lot of problems (since there is a lot of logic tailored for each specific `Target`). Furthermore, currently there is no such entity as a *proxy* target in meson, so `CustomTargetLinkTarget` will be considered a separate target from it's parent `CustomTarget`.

That said, my knowledge of meson architecture is pretty non-existent, so there might be a simple way to implement it.

**Comment by jpakkane** [comment]
There is also the case where people wanted to install files generated with custom target directly as manpages. We considered an `install_type` keyword that would tell Meson that the result needs to be installed as a man page. Maybe we can kill two birds with one stone and add a `treat_as` keyword argument that does both. Its value can be a string such as `static_library`, `executable` (the default), `manpage` etc.

**Comment by nirbheek** [comment]
> Maybe we can kill two birds with one stone and add a `treat_as` keyword argument that does both. Its value can be a string such as `static_library`, `executable` (the default), `manpage` etc.

What about the case when a custom target has multiple outputs, and each should be treated as something different? F.ex., generating headers (that will be installed to includedir) and also a C file which will be compiled? Or the output of `cargo` which can be one executable, one shared library, etc?

I guess we could have `treat_as:` take an array of strings for each output.

**Comment by TheQwertiest** [comment]
`treat_as` sounds nice, but I'm not sure where even to start... The only way I can think of (without rewriting half of meson) is to masquerade/replace the return value of `CustomTarget[i]` (which is currently `CustomTargetIndex`) depending on the value of `treat_as`, i.e. creating fake targets (`StaticLibrary`, `Executable`, dunno what to do about generated sources/headers and man pages though) from `CustomTargetIndex`. And I have no idea how to do the same for the `CustomTarget` itself (aside from dirty hacks like replacing the base/superclass).

**Comment by TheQwertiest** [comment]
PS: Any suggestions are welcome, since #4908 feature is a deal-breaker on adopting meson for us =(

**Comment by jpakkane** [comment]
Superseded by #5103.

**Pull request closed without merging**